File: 0257-Binary-Tree-Paths/0257.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> ans = new ArrayList<>();
        treePath(root, "", ans);
        return ans;
    }

    public void treePath(TreeNode node, String path, List<String> ans) {
        if (node == null)
            return;

        path += node.val;

        if (node.left == null && node.right == null) {
            ans.add(path);
            return;
        }

        if (node.left != null) {
            treePath(node.left, path + "->", ans);
        }

        if (node.right != null) {
            treePath(node.right, path + "->", ans);
        }
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

File: 0257-Binary-Tree-Paths/0257.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]: # type: ignore
        ans = []
        if not root: return ans

        def tree_path(node, path):
            path += str(node.val)

            if not node.left and not node.right:
                ans.append(path)
                return
            
            if node.left:
                tree_path(node.left, path + "->")
            
            if node.right:
                tree_path(node.right, path + "->")
        
        tree_path(root, "")
        return ans